### PR TITLE
add dependsOn option in field which is same as v4

### LINKED ISSUE #2231: "dependsOn" field prop in keystone 5
Where is the "dependsOn" property of fields  in list???

### PR DESCRIPTION
**Pull request by gautamsi**
related to #1355 
closes #2231 

this being v4 feature `dependsOn` which is useful for showing or hiding the component in admin ui based on other field values. 

example: 
dependsOn

following config used in basic test project

```js
keystone.createList('User', {
  fields: {
    name: { type: Text },
    email: { type: Text, isUnique: true, dependsOn: { name: 'email' } },
    dob: {
      type: CalendarDay,
      format: 'Do MMMM YYYY',
      yearRangeFrom: 1901,
      yearRangeTo: getYear(new Date()),
    },
    lastOnline: {
      type: DateTime,
      format: 'MM/DD/YYYY h:mm A',
      yearRangeFrom: 2013,
    },
    password: { type: Password },
    isAdmin: { type: Checkbox },
    favouritePosts: { type: Relationship, ref: 'Post', many: true, dependsOn: { isAdmin: true } },
    company: {
      type: Select,
      options: [
        { label: 'Thinkmill', value: 'thinkmill' },
        { label: 'Atlassian', value: 'atlassian' },
        { label: 'Thomas Walker Gelato', value: 'gelato' },
        { label: 'Cete, or Seat, or Attend ¯\\_(ツ)_/¯', value: 'cete' },
      ],
    },
    attachment: {
      type: File,
      adapter: fileAdapter,
      dependsOn: { $or: [{ company: ['cete', 'thinkmill'] }] },
    },
    color: { type: Color },
    website: { type: Url },
    ...(embedAdapter && { profile: { type: OEmbed, adapter: embedAdapter } }),
    ...(cloudinaryAdapter && { avatar: { type: CloudinaryImage, adapter: cloudinaryAdapter } }),
    ...(unsplash.accessKey && { favouriteImage: { type: Unsplash, ...unsplash } }),
  },
  labelResolver: item => `${item.name} <${item.email}>`,
  hooks: {
    validateInput: async ({ resolvedData, addValidationError }) => {
      if (resolvedData.name === 'Homer') {
        addValidationError('Sorry, no Homers allowed', { a: 1 }, { b: 2 });
      }
    },
  },
});

keystone.createList('Post', {
  fields: {
    name: { type: Text },
    slug: { type: Text },
    status: {
      type: Select,
      defaultValue: 'draft',
      options: [
        { label: 'Draft', value: 'draft' },
        { label: 'Published', value: 'published' },
      ],
    },
    author: {
      type: Relationship,
      ref: 'User',
    },
    categories: {
      type: Relationship,
      ref: 'PostCategory',
      many: true,
    },
    stars: { type: Stars, starCount: 5 },
    views: { type: Integer },
    price: { type: Decimal, symbol: '$' },
    currency: { type: Text, dependsOn: { $lt: { price: 3 } } },
    hero: { type: File, adapter: fileAdapter, dependsOn: { $gt: { price: 3 } } },
    markdownValue: { type: Markdown, dependsOn: { $gt: { price: 6 } } },
    fortyTwo: {
      type: Virtual,
      graphQLReturnType: `Int`,
      resolver: () => 42,
    },
    virtual: {
      type: Virtual,
      extendGraphQLTypes: [`type Movie { title: String, rating: Int }`],
      graphQLReturnType: `[Movie]`,
      graphQLReturnFragment: `{
        title
        rating
      }`,
      resolver: async () => {
        const data = [
          { title: 'A movie', rating: 2 },
          { title: 'Another movie', rating: 4 },
        ];
        return data.map(({ title, rating }) => ({ title, rating }));
      },
    },
    value: {
      type: Content,
      blocks: [
        ...(cloudinaryAdapter
          ? [[CloudinaryImage.blocks.image, { adapter: cloudinaryAdapter }]]
          : []),
        ...(embedAdapter ? [[OEmbed.blocks.oEmbed, { adapter: embedAdapter }]] : []),
        ...(unsplash.accessKey
          ? [[Unsplash.blocks.unsplashImage, { attribution: 'KeystoneJS', ...unsplash }]]
          : []),
        Content.blocks.blockquote,
        Content.blocks.orderedList,
        Content.blocks.unorderedList,
        Content.blocks.link,
        Content.blocks.heading,
      ],
    },
  },
  adminConfig: {
    defaultPageSize: 20,
    defaultColumns: 'name, status',
    defaultSort: 'name',
  },
  labelResolver: async (item, args, context, { schema }) => {
    if (item.author) {
      const query = `
        query getAuthor($authorId: ID!) {
          User(where: { id: $authorId }) {
            name
          }
        }
      `;
      const variables = { authorId: item.author.toString() };
      const { data } = await graphql(schema, query, null, context, variables);
      return `${item.name}${data.User && `(by ${data.User.name})`}`;
    } else {
      return item.name;
    }
  },
});

> changeset and possible test cases after core team evaluation
```

### CODE REVIEW

File: packages/utils/src/index.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ import pLazy from 'p-lazy';
 import pReflect from 'p-reflect';
 import isPromise from 'p-is-promise';
 import semver from 'semver';
+import ExMatch from '@mysupport/expression-match';
```
Comment (gautamsi) [inline packages/utils/src/index.js]: I had to fork this to resolve audit issue in lodash, nothing changes other than lodash version and cherry picking lodash functions.

**Comment by VinayaSathyanarayana** [comment]
This is really useful feature

**Comment by VinayaSathyanarayana** [comment]
While depends on can be an expression that is evaluated, I would also vote to make dependsOn a function that returns true / false.

**Comment by gautamsi** [comment]
> While depends on can be an expression that is evaluated, I would also vote to make dependsOn a function that returns true / false.

you can not transfer a function in node process to React app in browser, it is evaluated as an expression to overcome that issue.

**Comment by gautamsi** [comment]
fixed conflicts

**Comment by gautamsi** [comment]
fixed conflicts
maintainers: Please close this PR if not aligned with your vision, I would not want to keep it hanging without any future.

**Comment by VinayaSathyanarayana** [comment]
Let us get this into the master branch

**Comment by alobodig** [comment]
Is there any hope of getting this in master? I understand it was suggested to use the state machine feature to implement this, but that seems like it has a long way to go as well.

**Comment by mykiimike** [comment]
Please merge

**Comment by bladey** [comment]
Thanks for this PR @gautamsi.

**Keystone 5** has officially moved into active maintenance mode as we push towards the next major new version **Keystone Next**, you can find out more information about this transition [here](https://github.com/keystonejs/keystone/issues/4962).

In an effort to sustain the project going forward, we're cleaning up and closing old PRs such as this one.

However, you are welcome to create a PR under the [Keystone 5](https://github.com/keystonejs/keystone-5/pulls) repository.

Please let me know if you have any questions.

**Comment by emipc** [comment]
Did that ever reach or is it planned to reach Keystone Next? I asked about it in [stackoverflow](https://stackoverflow.com/questions/67984488/how-to-show-hide-fields-in-the-admin-ui-depending-on-the-value-from-another-fiel).

**Pull request closed without merging**